colossalai/communication/p2p.py

```diff
@@ -79,50 +79,108 @@ def _communicate(tensor_send_next: torch.Tensor = None,
-
-    if tensor_send_next is not None:
-        send_next_split = _get_tensor_shape(tensor_send_next.shape, scatter_gather_tensors)[1]
+            if isinstance(object_send_prev, torch.Tensor):
+                object_send_prev = split_tensor_into_1d_equal_chunks(object_send_prev)
+            else:
+                for tensor_send in object_send_prev:
+                    tensor_send = split_tensor_into_1d_equal_chunks(tensor_send)
+
+    if object_send_next is not None:
+        if isinstance(object_send_next, torch.Tensor):
+            send_next_split = _get_tensor_shape(object_send_next.shape, scatter_gather_tensors)[1]
+        else:
+            send_next_split = _get_tensor_shape(object_send_next[0].shape, scatter_gather_tensors)[1]
         if send_next_split:
-            tensor_send_next = split_tensor_into_1d_equal_chunks(tensor_send_next)
+            if isinstance(object_send_next, torch.Tensor):
+                object_send_next = split_tensor_into_1d_equal_chunks(object_send_next)
+            else:
+                for tensor_send in object_send_next:
+                    tensor_send = split_tensor_into_1d_equal_chunks(tensor_send)
 
     ops = []
-    if tensor_send_prev is not None:
-        send_prev_op = dist.P2POp(dist.isend, tensor_send_prev, prev_rank)
-        ops.append(send_prev_op)
+    if object_send_prev is not None:
+        if isinstance(object_send_prev, torch.Tensor):
+            send_prev_op = dist.P2POp(dist.isend, object_send_prev, prev_rank)
+            ops.append(send_prev_op)
+        else:
+            for tensor_send in object_send_prev:
+                send_prev_op = dist.P2POp(dist.isend, tensor_send, prev_rank)
+                ops.append(send_prev_op)
+
     if tensor_recv_prev is not None:
-        recv_prev_op = dist.P2POp(dist.irecv, tensor_recv_prev, prev_rank)
-        ops.append(recv_prev_op)
+        if isinstance(tensor_recv_prev, torch.Tensor):
+            recv_prev_op = dist.P2POp(dist.irecv, tensor_recv_prev, prev_rank)
+            ops.append(recv_prev_op)
+        else:
+            for tensor_recv in tensor_recv_prev:
+                recv_prev_op = dist.P2POp(dist.irecv, tensor_recv, prev_rank)
+                ops.append(recv_prev_op)
+
     if tensor_recv_next is not None:
-        recv_next_op = dist.P2POp(dist.irecv, tensor_recv_next, next_rank)
-        ops.append(recv_next_op)
-    if tensor_send_next is not None:
-        send_next_op = dist.P2POp(dist.isend, tensor_send_next, next_rank)
-        ops.append(send_next_op)
+        if isinstance(tensor_recv_next, torch.Tensor):
+            recv_next_op = dist.P2POp(dist.irecv, tensor_recv_next, next_rank)
+            ops.append(recv_next_op)
+        else:
+            for tensor_recv in tensor_recv_next:
+                recv_next_op = dist.P2POp(dist.irecv, tensor_recv, next_rank)
+                ops.append(recv_next_op)
+
+    if object_send_next is not None:
+        if isinstance(object_send_next, torch.Tensor):
```

Comment:
Will it be better to bandwidth utilisation if we concat tensors into a flatten tensor and set it out? But it will consumes more memory to prepare a flatten buff.

Reply:
I think concat tensors will achieve a better performance in communication, but it may not make a big difference in pipeline parallel scenario, because there are just several tensors to transport, not numbers of tiny tensors.